# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/base/nsFocusManager.cpp
# Commit: b3f0e5c7874d
# Full Hash: b3f0e5c7874dbd470d96e476e116740e4ca35925
# Author: David Zbarsky <dzbarsky@gmail.com>
# Date: 2012-08-06 14:40:43
# Regressor Bug: 779684
# File Overlap Count: 1
# Description:
#   Bug 779684: Create a stack-based nsFrameIterator class out of nsIFrameTraversal and nsIFrameEnumerator r=roc
# ==============================================================================

diff -r 4bc7154190dd -r b3f0e5c7874d dom/base/nsFocusManager.cpp
--- a/dom/base/nsFocusManager.cpp	Sun Aug 05 22:55:18 2012 -0400
+++ b/dom/base/nsFocusManager.cpp	Sun Aug 05 23:00:55 2012 -0400
@@ -32,7 +32,7 @@
 #include "nsLayoutUtils.h"
 #include "nsIPresShell.h"
 #include "nsIContentViewer.h"
-#include "nsFrameTraversal.h"
+#include "nsFrameIterator.h"
 #include "nsObjectFrame.h"
 #include "nsEventDispatcher.h"
 #include "nsEventStateManager.h"
@@ -2271,15 +2271,8 @@
         if (nodeValue.Length() == (PRUint32)startOffset && !isFormControl &&
             startContent != aDocument->GetRootElement()) {
           // Yes, indeed we were at the end of the last node
-          nsCOMPtr<nsIFrameEnumerator> frameTraversal;
-          nsresult rv = NS_NewFrameTraversal(getter_AddRefs(frameTraversal),
-                                             presContext, startFrame,
-                                             eLeaf,
-                                             false, // aVisual
-                                             false, // aLockInScrollView
-                                             true      // aFollowOOFs
-                                             );
-          NS_ENSURE_SUCCESS(rv, rv);
+          nsFrameIterator frameTraversal(presContext, startFrame,
+                                         eLeaf, FrameIteratorFlags::FLAG_FOLLOW_OUT_OF_FLOW);
 
           nsIFrame *newCaretFrame = nullptr;
           nsCOMPtr<nsIContent> newCaretContent = startContent;
@@ -2287,11 +2280,11 @@
           do {
             // Continue getting the next frame until the primary content for the frame
             // we are on changes - we don't want to be stuck in the same place
-            frameTraversal->Next();
-            newCaretFrame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
+            frameTraversal.Next();
+            newCaretFrame = static_cast<nsIFrame*>(frameTraversal.CurrentItem());
             if (nullptr == newCaretFrame)
               break;
-            newCaretContent = newCaretFrame->GetContent();            
+            newCaretContent = newCaretFrame->GetContent();
           } while (!newCaretContent || newCaretContent == startContent);
 
           if (newCaretFrame && newCaretContent) {
@@ -2704,21 +2697,14 @@
       continue;
     }
 
-    nsCOMPtr<nsIFrameEnumerator> frameTraversal;
-    nsresult rv = NS_NewFrameTraversal(getter_AddRefs(frameTraversal),
-                                       presContext, startFrame,
-                                       ePreOrder,
-                                       false, // aVisual
-                                       false, // aLockInScrollView
-                                       true      // aFollowOOFs
-                                       );
-    NS_ENSURE_SUCCESS(rv, rv);
+    nsFrameIterator frameTraversal(presContext, startFrame,
+                                   ePreOrder, FrameIteratorFlags::FLAG_FOLLOW_OUT_OF_FLOW);
 
     if (iterStartContent == aRootContent) {
       if (!aForward) {
-        frameTraversal->Last();
+        frameTraversal.Last();
       } else if (aRootContent->IsFocusable()) {
-        frameTraversal->Next();
+        frameTraversal.Next();
       }
     }
     else if (getNextFrame &&
@@ -2727,13 +2713,13 @@
       // Need to do special check in case we're in an imagemap which has multiple
       // content nodes per frame, so don't skip over the starting frame.
       if (aForward)
-        frameTraversal->Next();
+        frameTraversal.Next();
       else
-        frameTraversal->Prev();
+        frameTraversal.Prev();
     }
 
     // Walk frames to find something tabbable matching mCurrentTabIndex
-    nsIFrame* frame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
+    nsIFrame* frame = static_cast<nsIFrame*>(frameTraversal.CurrentItem());
     while (frame) {
       // TabIndex not set defaults to 0 for form elements, anchors and other
       // elements that are normally focusable. Tabindex defaults to -1
@@ -2803,10 +2789,10 @@
               Element* rootElement = subdoc->GetRootElement();
               nsIPresShell* subShell = subdoc->GetShell();
               if (rootElement && subShell) {
-                rv = GetNextTabbableContent(subShell, rootElement,
-                                            aOriginalStartContent, rootElement,
-                                            aForward, (aForward ? 1 : 0),
-                                            false, aResultContent);
+                nsresult rv = GetNextTabbableContent(subShell, rootElement,
+                                                     aOriginalStartContent, rootElement,
+                                                     aForward, (aForward ? 1 : 0),
+                                                     false, aResultContent);
                 NS_ENSURE_SUCCESS(rv, rv);
                 if (*aResultContent)
                   return NS_OK;
@@ -2850,10 +2836,10 @@
       // again.
       do {
         if (aForward)
-          frameTraversal->Next();
+          frameTraversal.Next();
         else
-          frameTraversal->Prev();
-        frame = static_cast<nsIFrame*>(frameTraversal->CurrentItem());
+          frameTraversal.Prev();
+        frame = static_cast<nsIFrame*>(frameTraversal.CurrentItem());
       } while (frame && frame->GetPrevContinuation());
     }
 